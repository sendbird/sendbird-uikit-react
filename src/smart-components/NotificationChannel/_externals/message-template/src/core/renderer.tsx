import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
