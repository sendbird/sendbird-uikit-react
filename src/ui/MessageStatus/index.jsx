import React from 'react';
import PropTypes from 'prop-types';

import './index.scss';
import Icon, { IconTypes, IconColors } from '../Icon';
import Label, { LabelColors, LabelTypography } from '../Label';
import Loader from '../Loader';

import {
  getMessageCreatedAt,
  getOutgoingMessageStates,
  isSentStatus,
} from '../../utils';

export const MessageStatusTypes = getOutgoingMessageStates();
export default function MessageStatus({
  className,
  message,
  status,
}) {
  const iconType = {
    [MessageStatusTypes.SENT]: IconTypes.DONE,
    [MessageStatusTypes.DELIVERED]: IconTypes.DONE_ALL,
    [MessageStatusTypes.READ]: IconTypes.DONE_ALL,
    [MessageStatusTypes.FAILED]: IconTypes.ERROR,
  };
  const iconColor = {
    [MessageStatusTypes.SENT]: IconColors.SENT,
    [MessageStatusTypes.DELIVERED]: IconColors.SENT,
    [MessageStatusTypes.READ]: IconColors.READ,
    [MessageStatusTypes.FAILED]: IconColors.ERROR,
  };

  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-message-status',
      ].join(' ')}
    >
      {(status === MessageStatusTypes.PENDING)
        ? (
          <Loader
            className="sendbird-message-status__icon"
            width="16px"
            height="16px"
          >
            <Icon
              type={IconTypes.SPINNER}
              fillColor={IconColors.PRIMARY}
              width="16px"
              height="16px"
            />
          </Loader>
        )
        : (
          <Icon
            className="sendbird-message-status__icon"
            type={iconType[status] || IconTypes.ERROR}
            fillColor={iconColor[status]}
            width="16px"
            height="16px"
          />
        )}
      {isSentStatus(status) && (
        <Label
          className="sendbird-message-status__text"
          type={LabelTypography.CAPTION_3}
          color={LabelColors.ONBACKGROUND_2}
        >
          {getMessageCreatedAt(message)}
        </Label>
      )}
    </div>
  );
}

MessageStatus.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  message: PropTypes.shape({
    createdAt: PropTypes.number,
    sender: PropTypes.shape({
      friendName: PropTypes.string,
      nickname: PropTypes.string,
      userId: PropTypes.string,
      profileUrl: PropTypes.string,
    }),
    sendingStatus: PropTypes.string,
  }),
  status: PropTypes.string,
};

MessageStatus.defaultProps = {
  className: '',
  message: null,
  status: '',
};
