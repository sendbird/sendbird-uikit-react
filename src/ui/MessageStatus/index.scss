@import '../../styles/variables';

.sendbird-message-status {
  position: relative;
  display: inline-flex;
  flex-direction: column;
  width: 100%;
  height: 100%;

  .sendbird-message-status__icon {
    position: absolute;
    top: 0px;
    right: 0px;
  }

  .sendbird-message-status__text {
    position: relative;
    margin-top: 18px;

    .sendbird-message-status__text__try-again {
      text-decoration: underline;
      &:hover {
        cursor: pointer;
      }
    }

    .sendbird-message-status__text__delete {
      text-decoration: underline;
      &:hover {
        cursor: pointer;
      }
    }
  }
}
