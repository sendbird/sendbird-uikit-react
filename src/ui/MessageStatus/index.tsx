import './index.scss';
import React, { useMemo } from 'react';
import format from 'date-fns/format';
import type { FileMessage, UserMessage } from '@sendbird/chat/message';
import type { GroupChannel } from '@sendbird/chat/groupChannel';

import Icon, { IconTypes, IconColors } from '../Icon';
import Label, { LabelColors, LabelTypography } from '../Label';
import Loader from '../Loader';

import {
  getOutgoingMessageState,
  getOutgoingMessageStates,
  isSentStatus,
} from '../../utils';
import { useLocalization } from '../../lib/LocalizationContext';

export const MessageStatusTypes = getOutgoingMessageStates();

interface MessageStatusProps {
  className?: string;
  message: UserMessage | FileMessage;
  channel: GroupChannel;
}

export default function MessageStatus({
  className,
  message,
  channel,
}: MessageStatusProps): React.ReactElement {
  const { dateLocale } = useLocalization();
  const status = useMemo(() => (
    getOutgoingMessageState(channel, message)
  ), [channel?.getUnreadMemberCount?.(message), channel?.getUndeliveredMemberCount?.(message)]);
  const hideMessageStatusIcon = channel?.isGroupChannel() && (
    (channel.isSuper || channel.isPublic || channel.isBroadcast)
    && !(status === MessageStatusTypes.PENDING || status === MessageStatusTypes.FAILED)
  );
  const iconType = {
    [MessageStatusTypes.SENT]: IconTypes.DONE,
    [MessageStatusTypes.DELIVERED]: IconTypes.DONE_ALL,
    [MessageStatusTypes.READ]: IconTypes.DONE_ALL,
    [MessageStatusTypes.FAILED]: IconTypes.ERROR,
  };
  const iconColor = {
    [MessageStatusTypes.SENT]: IconColors.SENT,
    [MessageStatusTypes.DELIVERED]: IconColors.SENT,
    [MessageStatusTypes.READ]: IconColors.READ,
    [MessageStatusTypes.FAILED]: IconColors.ERROR,
  };

  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-message-status',
      ].join(' ')}
    >
      {(status === MessageStatusTypes.PENDING) ? (
        <Loader
          className="sendbird-message-status__icon"
          width="16px"
          height="16px"
        >
          <Icon
            type={IconTypes.SPINNER}
            fillColor={IconColors.PRIMARY}
            width="16px"
            height="16px"
          />
        </Loader>
      ) : (
        <Icon
          className={`sendbird-message-status__icon ${hideMessageStatusIcon ? 'hide-icon' : ''}`}
          type={iconType[status] || IconTypes.ERROR}
          fillColor={iconColor[status]}
          width="16px"
          height="16px"
        />
      )}
      {isSentStatus(status) && (
        <Label
          className="sendbird-message-status__text"
          type={LabelTypography.CAPTION_3}
          color={LabelColors.ONBACKGROUND_2}
        >
          {format(message?.createdAt || 0, 'p', {
            locale: dateLocale,
          })}
        </Label>
      )}
    </div>
  );
}
