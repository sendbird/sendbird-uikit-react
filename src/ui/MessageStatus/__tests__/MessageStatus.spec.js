import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import MessageStatus, { MessageStatusTypes } from "../index";
import dummyMessage from '../messageDummyData.mock';

// mock date-fns to avoid problems from snapshot timestamping
// between testing in different locations
// ideally we want to mock date-fns globally - needs more research
jest.mock('date-fns/format', () => () => ('mock-date'));

describe('MessageStatus', () => {
  it('should contain className', function () {
    const text = "example-text";
    const component = shallow(<MessageStatus className={text} message={dummyMessage} />);

    expect(
      component.find(".sendbird-message-status").hasClass(text)
    ).toBe(true);
  });

  it('should do a snapshot test of the MessageStatus DOM', function () {
    const text = "example-text";
    const component = renderer.create(
      <MessageStatus className={text} message={dummyMessage} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should do a snapshot test of the failed MessageStatus DOM when isResendable: true', function () {
    const text = "example-text";
    const failedMsg = {
      ...dummyMessage,
      sendingStatus: 'failed',
      isResendable: () => { return true; },
    };
    const component = renderer.create(
      <MessageStatus className={text} message={failedMsg} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should do a snapshot test of the failed MessageStatus DOM when isResendable: false', function () {
    const text = "example-text";
    const failedMsg = {
      ...dummyMessage,
      sendingStatus: 'failed',
    };
    const component = renderer.create(
      <MessageStatus className={text} message={failedMsg} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
