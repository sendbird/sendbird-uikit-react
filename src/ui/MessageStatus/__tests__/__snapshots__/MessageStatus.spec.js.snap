// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageStatus should do a snapshot test of the MessageStatus DOM 1`] = `
<div
  className="example-text sendbird-message-status"
>
  <div
    className="sendbird-message-status__icon  sendbird-icon sendbird-icon-done sendbird-icon-color--sent"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    style={
      Object {
        "height": "16px",
        "minHeight": "16px",
        "minWidth": "16px",
        "width": "16px",
      }
    }
    tabIndex="0"
  >
    <test-file-stub />
  </div>
  <span
    className="sendbird-message-status__text sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-2"
  >
    mock-date
  </span>
</div>
`;

exports[`MessageStatus should do a snapshot test of the failed MessageStatus DOM when isResendable: false 1`] = `
<div
  className="example-text sendbird-message-status"
>
  <div
    className="sendbird-message-status__icon  sendbird-icon sendbird-icon-error sendbird-icon-color--error"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    style={
      Object {
        "height": "16px",
        "minHeight": "16px",
        "minWidth": "16px",
        "width": "16px",
      }
    }
    tabIndex="0"
  >
    <test-file-stub />
  </div>
</div>
`;

exports[`MessageStatus should do a snapshot test of the failed MessageStatus DOM when isResendable: true 1`] = `
<div
  className="example-text sendbird-message-status"
>
  <div
    className="sendbird-message-status__icon  sendbird-icon sendbird-icon-error sendbird-icon-color--error"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    style={
      Object {
        "height": "16px",
        "minHeight": "16px",
        "minWidth": "16px",
        "width": "16px",
      }
    }
    tabIndex="0"
  >
    <test-file-stub />
  </div>
</div>
`;
